src/gradient_descent.c

```
@@ -0,0 +1,380 @@
+                               int itMax) {
+  // Gradient of the function
+  gradient_real *gradient = malloc(n * sizeof(gradient_real));
+  gradient_real *nextGradient = malloc(n * sizeof(gradient_real));
+
+  // Conjugate gradient of the function
+  gradient_real *conjugate = malloc(n * sizeof(gradient_real));
+
+  // Compute initial value of function at the guess point
+  gradient_real value = (*func)(guess);
+
+  // Compute gradient of the function at the guess point
+  (*dfunc)(guess, gradient);
+
+  // Initialize the gradient and the conjugate
+  gradient_real temp;
+  for (int i = 0; i < n; i++) {
+    temp = -gradient[i];
+    gradient[i] = temp;
+    conjugate[i] = temp;
+  }
+
+  gradient_real min = 0.0;
+  for (int its = 0; its < itMax; its++) {
+    // Moves the guess to the minimum along the conjugate gradient direction
+    // Obtain the value of the function at that minimum
+    min = lineSearch(guess, conjugate, n, func, tol);
+
+    // Checks if we have reached a minimum
+    if (2.0 * fabs(min - value) <= tol * (fabs(min) + fabs(value) + EPS)) {
+      return min;
+    }
+
+    // Set the value of the function at the new guess point
+    value = min;
+
+    // Compute the gradient at the new guess point
+    (*dfunc)(guess, nextGradient);
+
+    // Compute the adjustment factor for the new conjugate
+    gradient_real dgg = 0.0;
+    gradient_real gg = 0.0;
+    for (int i = 0; i < n; i++) {
+      gg += gradient[i] * gradient[i];
+      dgg += (nextGradient[i] + gradient[i]) * nextGradient[i];
+    }
+    if (gg == 0.0) {
+      gg = EPS; // Prevent division by 0
+    }
+    gradient_real adjustement = dgg / gg;
+
+    // Set the new conjugate and gradient for the next iteration
+    for (int i = 0; i < n; i++) {
+      gradient[i] = -nextGradient[i];
+      conjugate[i] = gradient[i] + adjustement * conjugate[i];
+    }
+  }
+
+  // We should not reach here something went wrong
+  return min;
```

Again, should we return an error if we were not able to find a solution after the max number of iterations ?